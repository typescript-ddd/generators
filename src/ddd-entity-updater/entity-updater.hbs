import { DomainEventBus, EntityNotFoundError } from "@typescript-ddd/core";
import { {{pascal entityName}}Repository } from "../../infrastructure/persistence";
import { {{pascal entityName}} } from "../../domain";

/**
 * Represents {{noun entityName}} updater.
 */
export class {{pascal entityName}}Updater {
  /**
   * Initializes a new instance of {{noun entityName}} updater.
   * @param {{lbrace}}{{pascal entityName}}Repository{{rbrace}} repository - The {{natural entityName}} repository.
   * @param {DomainEventBus} eventBus - The domain event bus.
   */
  constructor(
    private readonly repository: {{pascal entityName}}Repository,
    private readonly eventBus: DomainEventBus
  ) {}

  /**
   * Updates the {{natural entityName}}.
   * @param {{lbrace}}Update{{pascal entityName}}Props{{rbrace}} props - the properties to update for the {{natural entityName}}.
   * @returns {Promise<void>}
   */
  public async update(props: Update{{pascal entityName}}Props): Promise<void> {
    const { id } = props;
    const entity = await this.repository.findById(id);
    if (!entity) {
      throw new EntityNotFoundError("{{pascal entityName}}", id);
    }
    {{pascal entityName}}.update(props);
    await this.repository.save(entity);
    this.eventBus.publish(entity.pullEvents());
  }
}
