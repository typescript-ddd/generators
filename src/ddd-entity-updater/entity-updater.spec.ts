import { EntityUpdaterTemplateData } from "./entity-updater.types";
import { generateEntityUpdater } from "./entity-updater.generator";

describe("EntityUpdater", () => {
    it("should render a updated entity template", () => {
        // Arrange
        const data: EntityUpdaterTemplateData = {
        entityName: "User"
        };
    
        // Act
        const result = generateEntityUpdater(data);
    
        // Assert
        expect(result).toMatchSnapshot();
    });
    });