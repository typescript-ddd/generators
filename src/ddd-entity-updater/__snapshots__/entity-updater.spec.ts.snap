// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityUpdater should render a updated entity template 1`] = `
"import { DomainEventBus, EntityNotFoundError } from "@typescript-ddd/core";
import { UserRepository } from "../../infrastructure/persistence";
import { User } from "../../domain";

/**
 * Represents a user updater.
 */
export class UserUpdater {
  /**
   * Initializes a new instance of a user updater.
   * @param {UserRepository} repository - The user repository.
   * @param {DomainEventBus} eventBus - The domain event bus.
   */
  constructor(
    private readonly repository: UserRepository,
    private readonly eventBus: DomainEventBus
  ) {}

  /**
   * Updates the user.
   * @param {UpdateUserProps} props - the properties to update for the user.
   * @returns {Promise<void>}
   */
  public async update(props: UpdateUserProps): Promise<void> {
    const { id } = props;
    const entity = await this.repository.findById(id);
    if (!entity) {
      throw new EntityNotFoundError("User", id);
    }
    User.update(props);
    await this.repository.save(entity);
    this.eventBus.publish(entity.pullEvents());
  }
}
"
`;
